// npm-http-server/components/IndexPage.js
'use strict';

exports.__esModule = true;

var _semver = require('semver');

var _semver2 = _interopRequireDefault(_semver);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _DirectoryListing = require('./DirectoryListing');

var _DirectoryListing2 = _interopRequireDefault(_DirectoryListing);

var _StyleUtils = require('../StyleUtils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var IndexPageStyle = (0, _StyleUtils.readCSS)(__dirname, 'IndexPage.css');
var IndexPageScript = '\nvar s = document.getElementById(\'version\'), v = s.value\ns.onchange = function () {\n  window.location.href = window.location.href.replace(\'@\' + v, \'@\' + s.value)\n}\n';

var byVersion = function byVersion(a, b) {
  return _semver2.default.lt(a, b) ? -1 : _semver2.default.gt(a, b) ? 1 : 0;
};

var IndexPage = function (_React$Component) {
  _inherits(IndexPage, _React$Component);

  function IndexPage() {
    _classCallCheck(this, IndexPage);

    return _possibleConstructorReturn(this, _React$Component.apply(this, arguments));
  }

  IndexPage.prototype.render = function render() {
    var _props = this.props,
        packageInfo = _props.packageInfo,
        version = _props.version,
        dir = _props.dir,
        entries = _props.entries;


    var versions = Object.keys(packageInfo.versions).sort(byVersion);
    var options = versions.map(function (v) {
      return _react2.default.createElement(
        'option',
        { key: v, value: v },
        packageInfo.name,
        '@',
        v
      );
    });

    return _react2.default.createElement(
      'html',
      null,
      _react2.default.createElement(
        'head',
        null,
        _react2.default.createElement('meta', { charSet: 'utf-8' }),
        _react2.default.createElement(
          'title',
          null,
          'Index of ',
          dir
        ),
        _react2.default.createElement('style', { dangerouslySetInnerHTML: { __html: IndexPageStyle } })
      ),
      _react2.default.createElement(
        'body',
        null,
        _react2.default.createElement(
          'div',
          { className: 'version-wrapper' },
          _react2.default.createElement(
            'select',
            { id: 'version', defaultValue: version },
            options
          )
        ),
        _react2.default.createElement(
          'h1',
          null,
          'Index of ',
          dir
        ),
        _react2.default.createElement('script', { dangerouslySetInnerHTML: { __html: IndexPageScript } }),
        _react2.default.createElement('hr', null),
        _react2.default.createElement(_DirectoryListing2.default, { dir: dir, entries: entries }),
        _react2.default.createElement('hr', null),
        _react2.default.createElement(
          'address',
          null,
          packageInfo.name,
          '@',
          version
        )
      )
    );
  };

  return IndexPage;
}(_react2.default.Component);

IndexPage.propTypes = {
  packageInfo: _react.PropTypes.object.isRequired,
  version: _react.PropTypes.string.isRequired,
  dir: _react.PropTypes.string.isRequired,
  entries: _react.PropTypes.array.isRequired
};
exports.default = IndexPage;
